src/part-guide/adv-async-await.md

```diff
@@ -2,26 +2,143 @@
+
+
+## Async blocks
+
+A regular block (`{ ... }`) groups code together in the source and creates a scope of encapsulation for names. At runtime, the block is executed in order and evaluates to the value of its last expression (or the unit type (`()`) if there is no trailing expression).
+
+Similarly to async functions, an async block is a deferred version of a regular block. An async block scopes code and names together, but at runtime it is not immediately executed and evaluates to a future. To execute the block and obtain the result, it must be `await`ed. E.g.,
+
+```rust,norun
+let s1 = {
+    let a = 42;
+    format!("The answer is {a}")
+};
+
+let s2 = async {
+    let q = question().await;
+    format!("The question is {q}")
+};
+```
+
+If we were to execute this snippet, `s1` would be a string which could be printed, but `s2` would be a future; `question()` would not have been called. To print `s2`, we first have to `s2.await`.
+
+An async block is the simplest way to start an async context and create a future. It is commonly used to create small futures which are only used in one place.
+
+Unfortunately, control flow with async blocks is a little quirky. Because an async block creates a future rather than straightforwardly executing, it behaves more like a function than a regular block with respect to control flow. `break` and `continue` cannot go 'through' an async block like they can with regular blocks; instead you have to use `return`:
+
+```rust,norun
+loop {
+    {
+        if ... {
+            // ok
+            continue;
+        }
+    }
+
+    async {
+        if ... {
+            // not ok
+            // continue;
+
+            // ok - continues with the next execution of the `loop`, though note that if there was
+            // code in the loop after the async block that would be executed.
+            return;
+        }
+    }.await
+}
+```
+
+To implement `break` you would need to test the value of the block (a common idiom is to use [`ControlFlow`](https://doc.rust-lang.org/std/ops/enum.ControlFlow.html) for the value of the block, which also allows use of `?`).
+
+Likewise, `?` inside an async block will terminate execution of the future in the presence of an error, causing the `await`ed block to take the value of the error, but won't exit the surrounding function (like `?` in a regular block would). You'll need another `?` after `await` for that:
+
+```rust,norun
+async {
+    let x = foo()?;   // This `?` only exits the async block, not the surrounding function.
+    consume(x);
+    Ok(())
+}.await?
+```
+
+Annoyingly, this often confuses the compiler since (unlike functions) the 'return' type of an async block is not explicitly stated. You'll probably need to add some type annotations on variables or use turbofished types to make this work, e.g., `Ok::<_, MyError>(())` instead of `Ok(())` in the above example.
```

Comment:
I think it'd be good to make this example compile and run as is, rather than saying you'll probably need to modify it. Or does this one work and we got lucky, but in general you'll probably need a turbofish?

Reply:
This is just a snippet, whether it compiles or not depends on the surrounding function (and also the type of `foo`, probably).